ReturnYouTubeDislike connection statistics. Show client toast messages on RYD connection errors.  High priority background threads.

LisoUseInAIKyrios opened the pull request:
I changed the background thread pool so all threads run at max priority.

This change seems to reduce the average time needed to load the video UI, as the RYD vote fetch now gets priority over other background threads (the fetch is blocking the UI, it definitely should get priority). 

Currently this thread pool is only used by RYD for the fetch votes calls.

oSumAtrIX commented:
Can you include benchmark results

LisoUseInAIKyrios commented:
Benchmarking is not precise since it's affected by whatever is going on in the background, how fast YouTube responds for each video, internet connection speed at that moment, etc, etc, etc.

Without patch.  Loading 20 random videos:

```
revanced: ReturnYouTubeDislike: UI thread paused for: 1071ms waiting for vote fetch to complete
revanced: ReturnYouTubeDislike: UI thread paused for: 1221ms waiting for vote fetch to complete
revanced: ReturnYouTubeDislike: UI thread paused for: 661ms waiting for vote fetch to complete
revanced: ReturnYouTubeDislike: UI thread paused for: 61ms waiting for vote fetch to complete
```

With patch.  Loading same 20 random videos:

```
revanced: ReturnYouTubeDislike: UI thread paused for: 762ms waiting for vote fetch to complete
revanced: ReturnYouTubeDislike: UI thread paused for: 159ms waiting for vote fetch to complete

```

oSumAtrIX commented:
Unsure if there is any improvement, both benchmarks show similar results.

LisoUseInAIKyrios commented:
without the patch, the UI waited 4 out of 20 times.

With the patch, it waited 2 out of 20 times.

oSumAtrIX commented:
Oh I see, i thought these were samples taken out of the 20.

oSumAtrIX commented on app/src/main/java/app/revanced/integrations/returnyoutubedislike/ReturnYouTubeDislike.java:
```diff
@@ -138,12 +138,19 @@ public static void onComponentCreated(Object conversionContext, AtomicReference<
+                votingData = fetchFuture.get(MILLISECONDS_TO_BLOCK_UI_WHILE_WAITING_FOR_DISLIKE_FETCH_TO_COMPLETE, TimeUnit.MILLISECONDS);
             } catch (TimeoutException e) {
                 LogHelper.printDebug(() -> "UI timed out waiting for dislike fetch to complete");
+                recordTimeUISpentWaitingForNetworkCall(fetchStartTime);
                 return;
             }
+            recordTimeUISpentWaitingForNetworkCall(fetchStartTime);
```
A try-catch-finally block can be used in order to call `recordTimeUISpentWaitingForNetworkCall` before return without using duplicate code.

LisoUseInAIKyrios commented:
I added logging to the API network calls, and added user viewable statistics if debugging is enabled:
https://freeimage.host/i/HoDNhKB

Also added a Toast if a connection times out, or if a rate limit is encountered.

oSumAtrIX commented on app/src/main/java/app/revanced/integrations/returnyoutubedislike/ReturnYouTubeDislike.java:
```diff
@@ -40,7 +40,7 @@ public class ReturnYouTubeDislike {
     private static final ExecutorService voteSerialExecutor = Executors.newSingleThreadExecutor();
 
-    // Must be volatile, since non-main threads read this field.
+    // Must be volatile, since this is read/wright from different threads
```
```suggestion
    // Must be volatile, since this is read/write from different threads
```

oSumAtrIX commented:
Thanks!
